src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php

```diff
@@ -59,14 +67,29 @@ protected function checkAuthentication(UserInterface $user, UsernamePasswordToke
                 throw new BadCredentialsException('The presented password is invalid.');
             }
 
-            $encoder = $this->encoderFactory->getEncoder($user);
+            // deprecated since Symfony 5.3
+            if ($this->hasherFactory instanceof EncoderFactoryInterface) {
+                $encoder = $this->hasherFactory->getEncoder($user);
+
+                if (!$encoder->isPasswordValid($user->getPassword(), $presentedPassword, $user->getSalt())) {
+                    throw new BadCredentialsException('The presented password is invalid.');
+                }
+
+                if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($encoder, 'needsRehash') && $encoder->needsRehash($user->getPassword())) {
+                    $this->userProvider->upgradePassword($user, $encoder->encodePassword($presentedPassword, $user->getSalt()));
+                }
+
+                return;
+            }
+
+            $hasher = $this->hasherFactory->getHasher($user);
 
-            if (!$encoder->isPasswordValid($user->getPassword(), $presentedPassword, $user->getSalt())) {
+            if (!$hasher->verify($user->getPassword(), $presentedPassword, $user->getSalt())) {
                 throw new BadCredentialsException('The presented password is invalid.');
             }
 
-            if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($encoder, 'needsRehash') && $encoder->needsRehash($user->getPassword())) {
-                $this->userProvider->upgradePassword($user, $encoder->encodePassword($presentedPassword, $user->getSalt()));
+            if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($hasher, 'needsRehash') && $hasher->needsRehash($user->getPassword())) {
```

Comment:
should we find a way to get rid of the method_exists check?

Reply:
I think we can already get rid of it, as `needsRehash()` is part of `PasswordHasherInterface`.